examples/dataflow-hashpipeline/hashpipeline/buffering.py

```diff
@@ -0,0 +1,143 @@
+import apache_beam as beam
+import time
+
+class PayloadFactory:
+  def __init__(self, filename, max_size):
+    self.payloads = []
+    self.max_size = max_size
+    self.filename = filename
+
+  def running_payload(self):
+    '''Returns a singleton instance of the current payload'''
+    raise NotImplementedError
+  
+  def flatten(self):
+    '''Adds current payload to payloads list and flattens'''
+    self.payloads.append(self.running_payload())
+    return [(self.filename, payload) for payload in self.payloads]
+  
+  def extend_running_payload(self, data):
+    '''Extends the current_singleton object until extending it further would pass max_size'''
+    raise NotImplementedError
+
+class FilePayloadFactory(PayloadFactory):
+  '''Payload factory for buffering lines in a file before sending batch requests to DLP'''
+  MAX_BYTE_SIZE_PER_REQUEST = 400000 # 0.4MB giving 100KB for metadata
+
+  def __init__(self, filename, max_size = MAX_BYTE_SIZE_PER_REQUEST):
+    super(FilePayloadFactory, self).__init__(filename, max_size) 
+    self._current = ""
+  
+  def running_payload(self):
+    return self._current
+  
+  # NOTE: This does not handle the case where a single line is longer than max_size
+  def extend_running_payload(self, data):
+    if len(self._current) + len(data) > self.max_size:
+      self.payloads.append(self._current)
+      self._current = ""
+    else:
+      self._current += data
+
+class ListPayloadFactory(PayloadFactory):
+  '''Payload factory for buffering SSN hash digests to batch get from Firestore'''
+  MAX_KEYS_PER_REQUEST = 1000 # 10MB max request size
+
+  def __init__(self, filename, max_size = MAX_KEYS_PER_REQUEST):
+    super(ListPayloadFactory, self).__init__(filename, max_size)
+    self._current = []
+  
+  def running_payload(self):
+    return self._current
+  
+  # NOTE: This does not handle the case where a single line is longer than max_size
+  def extend_running_payload(self, data):
+    if len(self._current) + len(data) > self.max_size:
+      self.payloads.append(self._current)
+      self._current = []
+    else:
+      self._current.append(data)
+
+
+class StatefulBufferingFn(beam.DoFn):
+  MAX_BUFFER_SIZE = 500 # 500 lines
+  BUFFER_STATE = BagStateSpec('buffer', TupleCoder((StrUtf8Coder(), StrUtf8Coder())))
+  COUNT_STATE = CombiningValueStateSpec('count', VarIntCoder(), CountCombineFn())
+  EXPIRY_TIMER = TimerSpec('expiry', TimeDomain.WATERMARK)
+  STALE_TIMER = TimerSpec('stale', TimeDomain.REAL_TIME)
+  MAX_BUFFER_DURATION = 1
+  ALLOWED_LATENESS = 5    
+  def __init__(self, payload_factory = FilePayloadFactory):
+    self.payload_factory = payload_factory
+ 
+  def process(self, element,
+    w=beam.DoFn.WindowParam,
+    buffer_state=beam.DoFn.StateParam(BUFFER_STATE),
+    count_state=beam.DoFn.StateParam(COUNT_STATE),
+    expiry_timer=beam.DoFn.TimerParam(EXPIRY_TIMER),
+    stale_timer=beam.DoFn.TimerParam(STALE_TIMER)) :
+    if count_state.read() == 0:
```

Comment:
Because you read this state every call it'd be great to defer / prefetch the reads.
I'm not sure this is possible in python SDK. please email the user@beam.apache.org mailing list.

context: https://beam.apache.org/documentation/programming-guide/#deferred-state-reads